--- python.m4	2005-07-01 14:29:14.000000000 -0700
+++ /Users/bmahler/git/mesos/m4/python.m4	2012-12-06 13:36:03.000000000 -0800
@@ -10,6 +10,11 @@
 # gives unlimited permission to copy and/or distribute it,
 # with or without modifications, as long as this notice is preserved.
 
+# Modified by Benjamin Mahler (benjamin.mahler@gmail.com).
+# I've modified this file to check for python2.6, of which automake <= 1.10
+# is oblivious. As such, we can kill this file entirely if or when we decide
+# to require automake >= 1.11.
+
 # AM_PATH_PYTHON([MINIMUM-VERSION], [ACTION-IF-FOUND], [ACTION-IF-NOT-FOUND])
 # ---------------------------------------------------------------------------
 # Adds support for distributing Python modules and packages.  To
@@ -39,8 +44,8 @@
   dnl $prefix/lib/site-python in 1.4 to $prefix/lib/python1.5/site-packages
   dnl in 1.5.
   m4_define_default([_AM_PYTHON_INTERPRETER_LIST],
-                    [python python2 python2.5 python2.4 python2.3 python2.2 dnl
-python2.1 python2.0 python1.6 python1.5])
+                    [python python2 python2.6 python2.5 python2.4 python2.3 dnl
+python2.2 python2.1 python2.0 python1.6 python1.5])
 
   m4_if([$1],[],[
     dnl No version check is needed.
