diff -rupN libev-3.8/ev.c libev-3.8-patched/ev.c
--- libev-3.8/ev.c	2009-07-29 02:36:02.000000000 -0700
+++ libev-3.8-patched/ev.c	2012-02-21 09:43:18.000000000 -0800
@@ -1894,10 +1894,10 @@ ev_default_loop (unsigned int flags)
       if (ev_backend (EV_A))
         {
 #ifndef _WIN32
-          ev_signal_init (&childev, childcb, SIGCHLD);
-          ev_set_priority (&childev, EV_MAXPRI);
-          ev_signal_start (EV_A_ &childev);
-          ev_unref (EV_A); /* child watcher should not keep loop alive */
+/*           ev_signal_init (&childev, childcb, SIGCHLD); */
+/*           ev_set_priority (&childev, EV_MAXPRI); */
+/*           ev_signal_start (EV_A_ &childev); */
+/*           ev_unref (EV_A); /\* child watcher should not keep loop alive *\/ */
 #endif
         }
       else
@@ -1917,8 +1917,8 @@ ev_default_destroy (void)
   ev_default_loop_ptr = 0;
 
 #ifndef _WIN32
-  ev_ref (EV_A); /* child watcher */
-  ev_signal_stop (EV_A_ &childev);
+/*   ev_ref (EV_A); /\* child watcher *\/ */
+/*   ev_signal_stop (EV_A_ &childev); */
 #endif
 
   loop_destroy (EV_A);
