diff --git a/conf/hadoop-env.sh b/conf/hadoop-env.sh
index ada5bef..76aaf48 100644
--- a/conf/hadoop-env.sh
+++ b/conf/hadoop-env.sh
@@ -9,7 +9,8 @@
 # export JAVA_HOME=/usr/lib/j2sdk1.5-sun
 
 # Extra Java CLASSPATH elements.  Optional.
-# export HADOOP_CLASSPATH=
+MESOS_CLASSPATH=${HADOOP_HOME}/contrib/mesos/hadoop-mesos-0.20.205.0.jar:${HADOOP_HOME}/build/contrib/mesos/hadoop-mesos-0.20.205.0.jar
+export HADOOP_CLASSPATH=${MESOS_CLASSPATH}:${HADOOP_CLASSPATH}
 
 # The maximum amount of heap to use, in MB. Default is 1000.
 # export HADOOP_HEAPSIZE=2000
