diff --git c/boost/config/compiler/visualc.hpp i/boost/config/compiler/visualc.hpp
index c0557de71..c06eaea45 100644
--- c/boost/config/compiler/visualc.hpp
+++ i/boost/config/compiler/visualc.hpp
@@ -319,13 +319,3 @@
 
 #  define BOOST_COMPILER "Microsoft Visual C++ version " BOOST_STRINGIZE(BOOST_COMPILER_VERSION)
 #endif
-
-//
-// last known and checked version is 19.10.25017 (VC++ 2017):
-#if (_MSC_VER > 1910)
-#  if defined(BOOST_ASSERT_CONFIG)
-#     error "Unknown compiler version - please run the configure tests and report the results"
-#  else
-#     pragma message("Unknown compiler version - please run the configure tests and report the results")
-#  endif
-#endif
